Fixed #25364 -- Added support for all selenium's browsers in tests.

akki opened the pull request:
https://code.djangoproject.com/ticket/25364

timgraham commented:
Did you test it?

For something like: `./tests/runtests.py admin_views --selenium=ff,gc --parallel=1`
I'm getting:
```
======================================================================
ERROR: test_cancel_delete_confirmation (admin_views.tests.SeleniumTests)
Cancelling the deletion of an object takes the user back one page.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/tests/admin_views/tests.py", line 4326, in test_cancel_delete_confirmation
    self.admin_login(username='super', password='secret', login_url=reverse('admin:index'))
  File "/home/tim/code/django/django/contrib/admin/tests.py", line 183, in admin_login
    self.selenium.get('%s%s' % (self.live_server_url, login_url))
AttributeError: 'SeleniumTests' object has no attribute 'selenium'
```

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -14,11 +15,39 @@ def process_response(self, request, response):
+class AdminSeleniumMetaClass(type):
+    def __new__(cls, name, bases, attrs):
+        """
+        Dynamically injects new methods for running tests in different browsers
```
We use PEP 0257 verb style for new code: "inject"

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -14,11 +15,39 @@ def process_response(self, request, response):
+                    for index, spec in enumerate(all_specs):
+                        new_func_name = '%s__%s' % (key, spec.replace('/', '_'))
+                        if index > 0:
+                            new_func = types.FunctionType(func.func_code,
```
Use hanging indent:
```
new_func = types.FunctionType(
    func.func_code,
    ...
)
```

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -26,26 +55,38 @@ class AdminSeleniumWebDriverTestCase(StaticLiveServerTestCase):
-        if not os.environ.get('DJANGO_SELENIUM_TESTS', False):
+    def _get_local_webdriver_class(self, spec):
+        """
+        Returns the webdriver for a local browser corresponding to the given
```
as above: Return ...

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -26,26 +55,38 @@ class AdminSeleniumWebDriverTestCase(StaticLiveServerTestCase):
+        if not os.environ.get('DJANGO_SELENIUM_SPECS', ''):
             raise SkipTest('Selenium tests not requested')
+        elif not hasattr(test_method, 'browser_spec'):
+            raise SkipTest('Please make sure your test class is decorated with @browserize')
```
inaccurate comment? I don't see any other mentions of "browserize".

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -26,26 +55,38 @@ class AdminSeleniumWebDriverTestCase(StaticLiveServerTestCase):
+        }
+        return import_string(browsers[spec])
+
+    def setUp(self):
```
Do you know if this has to be in `setUp` instead of `setUpClass` as it is now?

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -26,26 +55,38 @@ class AdminSeleniumWebDriverTestCase(StaticLiveServerTestCase):
-        if hasattr(cls, 'selenium'):
-            cls.selenium.quit()
-        super(AdminSeleniumWebDriverTestCase, cls)._tearDownClassInternal()
+            raise SkipTest(
```
I think this should be an error instead of skip.

akki replied:
Should it be a keyerror or something else ?

timgraham replied:
I'd suggest `RuntimeError`.

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -26,26 +55,38 @@ class AdminSeleniumWebDriverTestCase(StaticLiveServerTestCase):
+            raise SkipTest(
+                'Selenium specifications "%s" not valid or '
+                'corresponding WebDriver not installed: %s'
+                % (browser_spec, str(e)))
```
put closing parenthesis on the next line

timgraham commented on docs/internals/contributing/writing-code/unit-tests.txt:
```diff
@@ -121,12 +121,23 @@ Going beyond that, you can specify an individual test method like this::
-``--selenium`` option::
-
-   $ ./runtests.py --settings=test_sqlite --selenium admin_inlines
+Some admin tests require Selenium, a web browser and Python >= 2.6 to work via a
```
The mentions of "Python >= 2.6" and selenium (version > 2.38) are from an old version of these docs and now obsolete. Make the modifications without reintroducing what was removed in a3473454ada67c0a16efeabcb78950641d4ac93c.

timgraham commented on tests/admin_widgets/tests.py:
```diff
@@ -786,20 +785,11 @@ def test_calendar_show_date_from_input(self):
-@override_settings(ROOT_URLCONF='admin_widgets.urls')
-class DateTimePickerShortcutsSeleniumFirefoxTests(SeleniumDataMixin, AdminSeleniumWebDriverTestCase):
+@override_settings(ROOT_URLCONF='admin_widgets.urls',
+                   TIME_ZONE='Asia/Singapore')
```
move this to the previous line

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -26,26 +55,38 @@ class AdminSeleniumWebDriverTestCase(StaticLiveServerTestCase):
+        Returns the webdriver for a local browser corresponding to the given
+        browser spec.
+        """
+        browsers = {
```
Can you check if there's a way to pull this mapping from selenium or build it automatically so that if browsers are added there, we don't have to update Django?

akki commented:
Thanks @timgraham, I'll look into it.

akki commented:
`./tests/runtests.py admin_views --selenium=firefox --parallel=1` should work fine now. But it isn't supporting multiple browsers in the argument.

akki commented on django/contrib/admin/tests.py:
```diff
@@ -26,18 +57,37 @@ class AdminSeleniumWebDriverTestCase(StaticLiveServerTestCase):
+            'ie': 'selenium.webdriver.Ie',
+            'chrome': 'selenium.webdriver.Chrome',
+            'phantomjs': 'selenium.webdriver.PhantomJS',
+            'edge': 'selenium.webdriver.Edge',
```
Selenium now supports edge as well so I have updated the dictionary, if that's fine.

timgraham replied:
Did you check if there's a way to pull this mapping from selenium or build it automatically so that if browsers are added there, we don't have to update Django?

akki replied:
The next thing in my checklist. :)

akki commented on django/contrib/admin/tests.py:
```diff
@@ -26,18 +57,37 @@ class AdminSeleniumWebDriverTestCase(StaticLiveServerTestCase):
         except Exception as e:
-            raise SkipTest('Selenium webdriver "%s" not installed or not '
-                           'operational: %s' % (cls.webdriver_class, str(e)))
-        cls.selenium.implicitly_wait(10)
```
Don't know what good this is doing, it made impact neither on the testcases nor the time of execution.

timgraham replied:
It's needed. See 294d0d88158fa83eb095d663d054038ce9dfc3d4 and the corresponding ticket for an explanation.

akki commented on django/contrib/admin/tests.py:
```diff
@@ -26,17 +57,27 @@ class AdminSeleniumWebDriverTestCase(StaticLiveServerTestCase):
+        Return the webdriver for a local browser corresponding to the given
+        browser spec.
+        """
+        driver = 'selenium.webdriver.%s.webdriver.WebDriver' % spec
```
This will work only for selenium version greater than 2.45, most recent version is 2.51. Updated docs.

timgraham commented on docs/internals/contributing/writing-code/unit-tests.txt:
```diff
@@ -121,12 +121,22 @@ Going beyond that, you can specify an individual test method like this::
 --------------------------
 
-Some tests require Selenium and a Web browser (Firefox, Google Chrome, or
+Some tests require Selenium(>=2.45) and a Web browser (Firefox, Google Chrome, or
```
I'm not concerned about mentioning a minimum version of selenium because as fast as browsers are updated, older versions of selenium stop working fairly quickly in my experience. We can assume testers use the latest version.

timgraham replied:
Could you please remove "(>=2.45)" to address the above comment?

akki replied:
Sure. I thought it does not matter that I added it.

timgraham commented:
Please rebase your branch so it merges cleanly.

timgraham commented:
Looks like there's a problem where something like `./tests/runtests.py --selenium admin_widgets` will run all the tests. I think specifying `--selenium` without any values should raise an error. If we could automatically extract and output the possible browser values in that error, then we won't have to hardcode the browser list in the docs which will inevitably go stale.

This is coming along, keep up the good work.

akki commented on docs/internals/contributing/writing-code/unit-tests.txt:
```diff
@@ -121,12 +121,22 @@ Going beyond that, you can specify an individual test method like this::
+* ``opera``: Opera
+* ``ie``: Internet Explorer
+* ``chrome``: Google Chrome
+* ``phantomjs``: PhantomJS
```
Just for the sake of completeness, more browsers like edge, safare, etc. have been added by selenium. Should I list them all here ?

timgraham replied:
I suggested that `runtests.py --selenium` (with no spec) should raise an error with the possible choices. Then we can document that as the way to get possible values and don't have to keep the doc up to date as browsers come and go.

akki commented:
@timgraham Please take a final review.
As far as the task of getting the list of browsers is concerned, unfortunately I haven't been able to find anything very concrete yet. Maybe I should open a separate ticket for it ?

timgraham commented:
An option might be to enumerate the packages in `selenium.webdriver` and exclude the ones that we know aren't browsers (`common`, `remote`, `support`), however that seems a bit hacky. Would you like to raise the issue upstream (https://github.com/seleniumhq/selenium/issues) and see if the devs would be open to adding that list? If so, I agree we could defer that improvement to a separate ticket until the version of selenium with the improvement is released.

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -14,11 +15,41 @@ def process_response(self, request, response):
+                if isinstance(value, types.FunctionType) and key.startswith('test'):
+                    func = value
+                    for index, spec in enumerate(all_specs):
+                        new_func_name = '%s__%s' % (key, spec.replace('/', '_'))
```
Some comments in this method would be useful, e.g. why do we need to replace / with _ (I don't think any specs have / in them)?

akki replied:
This class is not doing much right now, I thought we wanted to keep it for building upon it in future. As a matter of fact, removing it completely won't break/change anything.
It was supposed to work when more than one browser were passed to the `--selenium` flag which as I mentioned earlier is not working.

timgraham replied:
Oh I see now, this is why `setUpClass` was moved to `setUp` in the original patch. Anyway, I think simpler is better for now so let's remove the metaclass. Maybe you can keep this around in a branch somewhere so we can get it back if we decide to at some point.

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -14,11 +15,41 @@ def process_response(self, request, response):
+        if all_specs:
+            for key, value in attrs.items():
+                if isinstance(value, types.FunctionType) and key.startswith('test'):
+                    func = value
```
I guess all "func" names could be called "method" since we're dealing with test methods, not functions.

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -14,11 +15,41 @@ def process_response(self, request, response):
+                                func.func_globals,
+                                new_func_name,
+                                func.func_defaults,
+                                func.func_closure
```
add a trailing comma (we do this so if more items are added later, we don't have to modify this line again)

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -26,17 +57,27 @@ class AdminSeleniumWebDriverTestCase(StaticLiveServerTestCase):
-    webdriver_class = 'selenium.webdriver.firefox.webdriver.WebDriver'
+
+    @classmethod
+    def _get_local_webdriver_class(self, spec):
```
this is now simple enough that I don't think we need a separate method anymore

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -26,17 +57,27 @@ class AdminSeleniumWebDriverTestCase(StaticLiveServerTestCase):
     def setUpClass(cls):
-        if not os.environ.get('DJANGO_SELENIUM_TESTS', False):
-            raise SkipTest('Selenium tests not requested')
+        browser_spec = os.environ.get('DJANGO_SELENIUM_SPECS', None)
```
`None` not needed (that's the default for `.get()`)

akki commented:
I am raising the issue upstream, keeping my fingers crossed. :)
For till then, I am raising a ticket on our own Trac to discuss what can be done.
Please let me know if anything else is required on this PR.

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -26,17 +57,27 @@ class AdminSeleniumWebDriverTestCase(StaticLiveServerTestCase):
+            cls.selenium = cls._get_local_webdriver_class(browser_spec)()
+        except:
+            msg = 'Selenium specifications not valid or'
+            msg += ' corresponding WebDriver not installed: %s' % browser_spec
```
combining this with the previous line will be fine

timgraham commented on docs/internals/contributing/writing-code/unit-tests.txt:
```diff
@@ -121,12 +121,22 @@ Going beyond that, you can specify an individual test method like this::
+For example, using Firefox (``--selenium=firefox``)::
 
-   $ ./runtests.py --settings=test_sqlite --selenium admin_inlines
+    $ ./runtests.py --selenium=firefox admin_views.tests.SeleniumTests
```
Let's not add "tests.SeleniumTests" -- the name or path might change (we need to reorganize the huge test file in admin_views/tests.py) and won't be necessary when https://code.djangoproject.com/ticket/25653 is completed.

timgraham commented on docs/internals/contributing/writing-code/unit-tests.txt:
```diff
@@ -121,12 +121,22 @@ Going beyond that, you can specify an individual test method like this::
 install the selenium_ package into your Python path and run the tests with the
-``--selenium`` option::
+``--selenium=<BROWSER>`` option.
+For example, using Firefox (``--selenium=firefox``)::
```
flow this to the previous line

timgraham commented:
Check `tests/view_tests/tests/test_i18n.py` for another selenium test.

timgraham commented:
I don't think we need another Trac ticket yet.
Also, if we can't support multiple browser specs, I think that'll be a blocker to merge this ... at least until https://code.djangoproject.com/ticket/25653 is completed. Otherwise we have to run the entire test suite twice on our CI server (we currently test with firefox and chrome there) which will make a slow build even slower. I'll try to get some movement on that ticket though -- https://github.com/django/django/pull/5585 is the first part.

akki commented:
@timgraham I was working on the support for multi browser thing and have been able to do it. Just polishing it up. Will push it soon.

akki commented:
Hi @timgraham
I have been able to take in multiple arguments for most of the test classes but am facing some issue while doing it for classes which are inheriting multiple classes like [this one](https://github.com/django/django/pull/6099/files?diff=split#diff-8d7bf730629e571af360fcf0cf59165dR854).
I am not able to call the setUp function of `AdminSeleniumWebDriverTestCase` class as it is being `SeleniumDataMixin`'s `setUp` function overrides it. Could you please help me out ?
One solution could be to remove the `SeleniumDataMixin` class as is the case in other files.

timgraham commented:
I think as long as you have all the needed `super()` calls, it should work.

akki commented:
Please have a look.

charettes commented:
Since all subclasses of `admin_widgets.tests.SeleniumDataMixin` also subclass `admin_widgets.tests.AdminSeleniumWebDriverTestCase` I would declare a `AdminWidgetSeleniumWebDriverTestCase` class to avoid the `setUp`'s `super()` quirks:

```python
@override_settings(ROOT_URLCONF='admin_widgets.urls')
class AdminWidgetSeleniumWebDriverTestCase(AdminSeleniumWebDriverTestCase):
    available_apps = ['admin_widgets'] + AdminSeleniumWebDriverTestCase.available_apps

    def setUp(self):
        super(AdminWidgetSeleniumWebDriverTestCase, self).setUp()
        self.u1 = User.objects.create_superuser(username='super', password='secret', email='super@example.com')

...

class DateTimePickerSeleniumFirefoxTests(AdminWidgetSeleniumWebDriverTestCase):
    ....
```

akki commented:
Does this look good ?
Thanks @charettes for the much cleaner solution. :)

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -26,26 +56,35 @@ class AdminSeleniumWebDriverTestCase(StaticLiveServerTestCase):
-            raise SkipTest('Selenium webdriver "%s" not installed or not '
-                           'operational: %s' % (cls.webdriver_class, str(e)))
-        cls.selenium.implicitly_wait(10)
+        '''
```
We use double quotes for all docstrings.

timgraham commented on django/contrib/admin/tests.py:
```diff
@@ -26,26 +56,35 @@ class AdminSeleniumWebDriverTestCase(StaticLiveServerTestCase):
+        '''
+        Validate the browsers requested and raise appropriate exception.
+        '''
+        browsers = os.environ.get('DJANGO_SELENIUM_SPECS', '').split(',')
```
I'd do this validation in `runtests.py` -- no need to repeat it for every test classs.

timgraham commented on tests/view_tests/tests/test_i18n.py:
```diff
@@ -256,34 +257,68 @@ def test_i18n_with_locale_paths(self):
 
 
-skip_selenium = not os.environ.get('DJANGO_SELENIUM_TESTS', False)
+class JavascriptI18nTestsMetaClass(type):
```
Let's avoiding duplicating this metaclass and put it in a new file: `django/test/selenium.py`

timgraham commented on tests/view_tests/tests/test_i18n.py:
```diff
@@ -256,34 +257,68 @@ def test_i18n_with_locale_paths(self):
-    def tearDownClass(cls):
-        cls.selenium.quit()
-        super(JavascriptI18nTests, cls).tearDownClass()
+    def setUp(self):
```
Similarly, let's see if setUp/tearDown can be moved to a shared mixin.

timgraham commented on tests/view_tests/tests/test_i18n.py:
```diff
@@ -256,34 +257,68 @@ def test_i18n_with_locale_paths(self):
+        self.selenium.implicitly_wait(10)
+
+    def tearDown(self):
+        if hasattr(self, 'selenium'):
```
why hasattr?

timgraham commented on tests/admin_widgets/tests.py:
```diff
@@ -619,10 +619,16 @@ def test_on_delete_cascade_rel_cant_delete_related(self):
 
 
 @override_settings(ROOT_URLCONF='admin_widgets.urls')
-class DateTimePickerSeleniumFirefoxTests(SeleniumDataMixin, AdminSeleniumWebDriverTestCase):
```
`SeleniumDataMixin` is no longer used I think.

timgraham commented:
I'm getting `AttributeError: 'function' object has no attribute 'browser_spec'` with `./tests/runtests.py --selenium=chrome,firefox`

akki commented:
@timgraham 
It must have creeped in when I made a one-liner change after running the whole test suite. Sorry about that.
Everything should work fine now.

akki commented on tests/runtests.py:
```diff
@@ -234,6 +235,22 @@ def actual_test_processes(parallel):
+            # Create a class instance to see if everything is fine.
+            import_string('selenium.webdriver.%s.webdriver.WebDriver' % browser)().quit()
+        except:
+            msg = 'Selenium specifications not valid or corresponding WebDriver not installed: %s' % browser
```
The string `'selenium.webdriver.%s.webdriver.WebDriver' % browser_spec` is being hard-coded at a lot of places now. I think it would be better to make a function in some file which would return this string or maybe the driver itself. A `get_selenium_driver()` in `django/test/selenium.py`? `django/test/utils`?

timgraham replied:
sure, in `selenium.py` sounds okay.

akki commented on tests/admin_widgets/tests.py:
```diff
@@ -619,10 +614,16 @@ def test_on_delete_cascade_rel_cant_delete_related(self):
 
 @override_settings(ROOT_URLCONF='admin_widgets.urls')
-class DateTimePickerSeleniumFirefoxTests(SeleniumDataMixin, AdminSeleniumWebDriverTestCase):
+class AdminWidgetSeleniumWebDriverTestCase(AdminSeleniumWebDriverTestCase):
```
I think we can omit `"WebDriver"` from this class name.

timgraham replied:
sure

akki commented on django/contrib/admin/tests.py:
```diff
@@ -17,7 +14,7 @@ def process_response(self, request, response):
     MIDDLEWARE_CLASSES={'append': 'django.contrib.admin.tests.CSPMiddleware'},
 )
-class AdminSeleniumWebDriverTestCase(StaticLiveServerTestCase):
+class AdminSeleniumWebDriverTestCase(SeleniumTestCaseMixin, StaticLiveServerTestCase):
```
I think we can omit `"WebDriver"` from this class name.

timgraham replied:
sure

akki commented:
@timgraham Is anything else needed to be done in this patch.
Also I have added notes on some diffs, please let me know your views on them.

charettes commented on tests/runtests.py:
```diff
@@ -233,6 +234,22 @@ def actual_test_processes(parallel):
+        try:
+            # Create a class instance to see if everything is fine.
+            import_string('selenium.webdriver.%s.webdriver.WebDriver' % browser)().quit()
+        except:
```
Is there a reason for using a bare `except` here? If it's the case please add a comment else use `except ImportError`.

akki replied:
It's actually not always an `ImportError`. Sometimes this line would also give `Selenium`'s `WebDriverError` for example when you give a valid browser name (like chrome) but don't have the corresponding drivers installed.
The error message thus provided has two parts "Selenium specifications not valid"(for `ImportError`) and "corresponding WebDriver not installed" for `WebDriverError`.

I can break this into something like:
```python
except ImportError:
  msg = 'Selenium specifications not valid.'
except:
  msg = 'Corresponding WebDriver not installed.'
```
if that is more perferable

charettes replied:
As long as you use `except Exception` and not a bare `except` this should be good.

timgraham commented:
Could you think about this in light of https://code.djangoproject.com/ticket/25653 and d4dc775620fc57e962165eab98a77264e3dd16b2? I think adding a commit that tags the selenium test cases and appends "selenium" to tags in `runtests.py` should be sufficient but I just want to make sure this approach and the solution for that ticket are compatible. Thanks!

akki commented:
Yes, I agree that such a commit would be sufficient. The `--selenium` flag would decide if selenium tests have to run or not. `--tags` would decide which tests have to be run. To put it simple things would work like this:

```
./runtests.py                                    -> Run all tests other than selenium.
./runtests.py --selenium=firefox,chrome          -> Run all tests.
./runtests.py --tags=selenium --selenium=firefox -> Run only selenium tests.
```

So I believe they are compatible.

charettes commented on tests/runtests.py:
```diff
@@ -391,8 +408,8 @@ def paired_tests(paired_test, options, test_labels, parallel):
     parser.add_argument(
-        '--selenium', action='store_true', dest='selenium', default=False,
-        help='Run the Selenium tests as well (if Selenium is installed).')
+        '--selenium', dest='selenium', default=None,
```
What about using `action='append'` instead of spliting by comma?

akki replied:
I feel `selenium=firefox,chrome` would be more convenient than `--selenium=firefox --selenium=chrome`.

akki commented:
Done.

akki commented:
@timgraham 
I tried things out [here - d4dc77](https://github.com/akki/django/commit/9be5488c6620f04136798beab95ffd510cff74e0) regarding the compatibility concerns of the two patches you mentioned earlier.
Everything works fine and as mentioned by me in my earlier comment. I can send a PR in favour of [#25653](https://code.djangoproject.com/ticket/25653) once this current patch gets merged.

timgraham commented:
Still not working for me:
```
$ ./tests/runtests.py --selenium=firefox admin_views
======================================================================
ERROR: test_prepopulated_fields (admin_views.tests.SeleniumTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/python3.5.1/lib/python3.5/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/opt/python3.5.1/lib/python3.5/unittest/case.py", line 596, in run
    self.setUp()
  File "/home/tim/code/django/tests/admin_views/tests.py", line 4127, in setUp
    super(SeleniumTests, self).setUp()
  File "/home/tim/code/django/django/test/selenium.py", line 46, in setUp
    browser_spec = test_method.browser_spec
AttributeError: 'function' object has no attribute 'browser_spec'
```
You can add the tag commit as a separate one on this pull request. It would be convenient if `--selenium` automatically set `--tags=selenium` in `runtests.py`.

akki commented:
I am unable to reproduce this error, it is working fine on my end. I have forced pushed quite a times so please ensure that you have the latest version.
Is the error being shown only for `admin_views.tests.SeleniumTests.test_prepopulated_fields` or all other functions as well ?

timgraham commented on django/test/selenium.py:
```diff
@@ -0,0 +1,53 @@
+@skipIf(skip_selenium, 'Selenium tests not requested.')
+class SeleniumTestCaseMixin(object):
+
+    __metaclass__ = SeleniumMetaClass
```
Ah, need to use https://pythonhosted.org/six/#six.add_metaclass to fix it on Python 3 (that's why I had problems).

timgraham commented:
buildbot, test on selenium.

timgraham commented:
buildbot, test on selenium.

timgraham commented:
Not sure why the tests don't work on Jenkins. Will investigate later.

akki commented:
Okay but that should not have happened given that they used to run for firefox, chrome and ie, all three till now. We have not made any changes to the way we use `selenium` library and it's drivers as such.

timgraham commented:
There's some problem in the initialization:
```
Traceback (most recent call last):
  File "./runtests.py", line 246, in validate_selenium
    import_string('selenium.webdriver.%s.webdriver.WebDriver' % browser)().quit()
  File "/home/jenkins/workspace/pull-requests-selenium/database/sqlite3/label/trusty-pr/python/python3.4/tests/.env/lib/python3.4/site-packages/selenium/webdriver/firefox/webdriver.py", line 78, in __init__
    self.binary, timeout)
  File "/home/jenkins/workspace/pull-requests-selenium/database/sqlite3/label/trusty-pr/python/python3.4/tests/.env/lib/python3.4/site-packages/selenium/webdriver/firefox/extension_connection.py", line 51, in __init__
    self.binary.launch_browser(self.profile, timeout=timeout)
  File "/home/jenkins/workspace/pull-requests-selenium/database/sqlite3/label/trusty-pr/python/python3.4/tests/.env/lib/python3.4/site-packages/selenium/webdriver/firefox/firefox_binary.py", line 68, in launch_browser
    self._wait_until_connectable(timeout=timeout)
  File "/home/jenkins/workspace/pull-requests-selenium/database/sqlite3/label/trusty-pr/python/python3.4/tests/.env/lib/python3.4/site-packages/selenium/webdriver/firefox/firefox_binary.py", line 98, in _wait_until_connectable
    raise WebDriverException("The browser appears to have exited "
selenium.common.exceptions.WebDriverException: Message: The browser appears to have exited before we could connect. If you specified a log_file in the FirefoxBinary constructor, check it for details.
```

akki commented:
Okay, it might be because of this line https://github.com/django/django/pull/6099/files#diff-bf0d7f89b37ba018fb5c5b2833b37dd3R246
where I am initialising the class and using `quit()` in the same go. Do you think that might be the issue ?

timgraham commented:
Yes, something related to that.

akki commented:
buildbot, test on selenium.

timgraham commented:
Breaking it up into two lines is logically as doing it in one line.

akki commented:
buildbot, test on selenium.

akki commented:
Yes, I realised that later, I had mistaken it for something else.
Do we have a log_file as suggested in the error. It might have something we can use.

akki commented:
buildbot, test on selenium.

timgraham commented:
Please rebase your branch. I merged the other ticket in 6670da75ff8a59b2ec0b465846e3f76aab9155b2.

akki commented:
7buildbot, test on selenium.

akki commented:
buildbot, test on selenium.

akki commented:
buildbot, test on selenium.

akki commented:
buildbot, test on selenium.

akki commented:
@timgraham 
I ran the test suite on machines on Mac, Ubuntu and Windows on both Python 2 and 3, and it passes on all of them. Since I cannot reproduce the bug, I am unable to debug it on my end. Please try to run the tests on your computer as well to check if the tests pass or not. I reckon that the problem has something to do with Jenkins.

timgraham commented on tests/runtests.py:
```diff
@@ -233,6 +234,22 @@ def actual_test_processes(parallel):
+    for browser in browsers:
+        try:
+            # Create a class instance to see if everything is fine.
+            import_string('selenium.webdriver.%s.webdriver.WebDriver' % browser)().quit()
```
If we remove this will the tests run on Jenkins? It might be fine.

akki replied:
Removing this would raise a browser error on each method, when an invalid browser is provided. Right now, it is being raised only once in the beginning.
I don't think removing this would resolve the Jenkins error though.

timgraham replied:
Please give it a try, I'd like to see the new error.

akki replied:
Sure.
I have an exam in a few hours, so I am planning to study a bit now :P
Will do this tomorrow.

timgraham commented on tests/runtests.py:
```diff
@@ -233,6 +234,22 @@ def actual_test_processes(parallel):
+    """
+    Validate the browsers requested and raise appropriate exception.
+    """
+    # Remove any browser repition.
```
Don't think we need to worry about duplicates.

akki replied:
If the alphabetically first browser is duplicated, we would have the tests running for it twice. It is not a big thing though. That's the only problem duplicate browsers will bring up.

akki commented:
buildbot, test on selenium.

timgraham commented:
I'm not sure why Jenkins fails. I'll try to investigate sometime.

akki commented:
Okay. Thanks.

akki commented:
@timgraham 
Some facts that might be useful when you investigate this issue -
1) The last selenium test that actually ran and passed was [#56](http://djangoci.com/job/pull-requests-selenium/56/) which was for the [PR-6077](https://github.com/django/django/pull/6077) and not [#61](http://djangoci.com/job/pull-requests-selenium/61/).
2) The reason [#61](http://djangoci.com/job/pull-requests-selenium/61/) requested by [PR- 6165](https://github.com/django/django/pull/6165) had passed was because all tests had been skipped by Jenkins (because currently the logic is to skip any test which is not able to import the selenium webdriver). The result was **Ran 0 tests, skipped 31**.
3) This PR fails the tests because I have changed that [current logic](https://github.com/django/django/blob/master/django/contrib/admin/tests.py#L36) to raise an error instead.

So basically, any selenium test that has passed after [#56](http://djangoci.com/job/pull-requests-selenium/56/) is because all tests were skipped. The test that had run after that, i.e. [#57](http://djangoci.com/job/pull-requests-selenium/57/) was none other than the one in which the `EXTRA_RUNTESTS_ARGS` variable had been changed.

Conclusions:
1) Now I am quite sure that something with Jenkin's environment has changed and the recent failures of this PR have nothing to do with the patch.
2) #6165 might have broken on some test because it was never actually tested. Fortunately, it didn't (I ran the test suite on my machine and they pass).
3) Not to say, this also means that Jenkins is not a reliable place for running selenium tests right now, it will always give a green light provided that the "current logic" is there.

timgraham commented:
buildbot, test on selenium.

timgraham commented:
Good find. I found at least one problem with the Jenkins configuration and am working on it now.

timgraham commented:
buildbot, test on selenium.

timgraham commented:
buildbot, test on selenium.

akki commented:
buildbot, test on selenium.

timgraham commented:
Multiple browsers aren't working on Python 3. Zero tests are run for `--selenium=firefox,chrome`.

akki commented:
buildbot, test on selenium.

akki commented:
buildbot, test on selenium.

akki commented:
@timgraham - For some reason, the [selenium test for python 3](http://djangoci.com/view/PR%20Builders/job/pull-requests-selenium/database=sqlite3,label=trusty-pr,python=python3.5/77/console) is stuck on running the migration _Applying sites_framework.0001_initial..._.
Can you please abort it ?

timgraham commented:
Here are a few edits: http://dpaste.com/031M2DG
Otherwise this looks good to go, but if we can improve it to launch the browser once per class as described on the ticket, that will be useful and might even reduce the likelihood of Jenkins hanging. Still, the build does pass now on my local test Jenkins so that's a good sign.

MoritzS commented:
I guess we can address that in a separate ticket if the selenium tests on the CI still run in an acceptable time.

akki commented:
Updated -
1) The changes suggested http://dpaste.com/031M2DG except
```bash
-@override_settings(ROOT_URLCONF='admin_changelist.urls')
+@override_settings(ROOT_URLCONF="admin_changelist.urls")
```
Single inverted commas are being used in all files for this so I didn't change this one either. Is there some specific reason for changing it here ?

2) Added `validate_selenium` function so that a browser error is not raised on each method, when an invalid browser is provided. Now, it is being raised only once in the beginning.

3) Squashed and rebased.

timgraham commented:
It's better not to make unrelated cosmetic cleanups like change quote styles in a feature commit like this - that's why I changed it back in my edit.

I think we should investigate launching the browser once per class before merging -- otherwise it seems like we lose the ability to run selenium tests on Jenkins due to the hanging issue.

MoritzS commented:
@akki, do you want to work on launching the browser once per class yourself?
I could also open a PR and base it on yours.

akki commented:
I'll like to try it myself, if that would not be an issue. I'll give my 100% and try to make it as early as possible. (This is my first big patch for Django so I am quite excited about doing it :) )

MoritzS commented:
Yes of course, go ahead!

charettes commented on django/test/selenium.py:
```diff
@@ -0,0 +1,60 @@
+        """
+        browsers = os.environ.get('DJANGO_SELENIUM_SPECS', '').split(',')
+        if browsers[0]:
+            for key in list(attrs.keys()):
```
Use `attrs.items()` as you are accessing `attrs[key]` anyway.

charettes commented on django/test/selenium.py:
```diff
@@ -0,0 +1,60 @@
+        browsers = os.environ.get('DJANGO_SELENIUM_SPECS', '').split(',')
+        if browsers[0]:
+            for key in list(attrs.keys()):
+                if isinstance(attrs[key], types.FunctionType) and key.startswith('test'):
```
Not sure we should be checking for `types.FunctionType` as some decorator might be `__call__()`able class instances. I think `callable(attr)` would be more appropriate.

charettes commented on django/test/selenium.py:
```diff
@@ -0,0 +1,60 @@
+@tag('selenium')
+@add_metaclass(SeleniumMetaClass)
+@skipIf(skip_selenium, 'Selenium tests not requested.')
+class SeleniumTestCaseMixin(object):
```
Is there a reason this isn't simply a subclass of `LiveServerTestCase`?

charettes commented on django/test/selenium.py:
```diff
@@ -0,0 +1,60 @@
+                if isinstance(attrs[key], types.FunctionType) and key.startswith('test'):
+                    method = attrs[key]
+                    for index, spec in enumerate(browsers):
+                        # Create new methods for browsers other than first one.
```
I think creating a method for each browsers would make more sense as the method name makes it clear on which of the specified browser failed.

charettes commented on django/test/selenium.py:
```diff
@@ -0,0 +1,60 @@
+                                str(method_name),
+                                get_function_defaults(method),
+                                get_function_closure(method),
+                            )
```
A simple closure should be sufficient here.

```python
from functools import wraps

def create_browser_test(browser, test):
    @wraps(test)
    def browser_test(self, *args, **kwargs):
        self.browser = browser
        return test(self, *args, **kwargs)
    return browser_test
```

charettes commented on django/test/selenium.py:
```diff
@@ -0,0 +1,60 @@
+
+@tag('selenium')
+@add_metaclass(SeleniumMetaClass)
+@skipIf(skip_selenium, 'Selenium tests not requested.')
```
If we remove the original `test_*` methods and only add them based on `browsers` we can remove this `skipIf` as the test won't have any methods anyway.

charettes commented on django/test/selenium.py:
```diff
@@ -0,0 +1,60 @@
+        Dynamically inject new methods for running tests in different browsers
+        when multiple browser specs are provided (e.g. --selenium=firefox,chrome).
+        """
+        browsers = os.environ.get('DJANGO_SELENIUM_SPECS', '').split(',')
```
What about defining a `browsers` class attribute and make `runtests` set `SeleniumMetaClass.browsers = os.environ.get('DJANGO_SELENIUM_SPECS', '').split(',')` instead?

This would allow users to either set the browsers to be tested per test class or create a subclass that can be used project wide.

charettes commented on django/test/selenium.py:
```diff
@@ -0,0 +1,60 @@
+        test_method = getattr(self, self._testMethodName)
+        browser_spec = test_method.browser_spec
+        self.selenium = import_string('selenium.webdriver.%s.webdriver.WebDriver' % browser_spec)()
+        self.selenium.implicitly_wait(10)
```
Here's what I suggest you do to avoid creating a WebDriver per test.

In `setupClass` create a `dict` mapping tested browser names to `WebDriver` instances. Define a `selenium` `cached_property` that retrieves the test specific driver from the mapping you created in `setupClass` by using `self.browser` (it should be set by the `create_browser_test` closure defined in a previous comment). Make sure to loop `quit()` all drivers in `tearDownClass()`.

MoritzS replied:
If we do this that way a test class would run as follows:
```
start browser 1
start browser 2
run test_x on browser 1
run test_x on browser 2
run test_y on browser 1
run test_y on browser 2
stop browser 1
stop browser 2
```
So all browser you are currently running tests on are started at the same time and will be active throughout the test case. I think it would be better to have the tests run as they do now, i.e.:
```
start browser 1
run test_x on browser 1
run test_y on browser 1
stop browser 1
start browser 2
run test_x on browser 2
run test_y on browser 2
stop browser 2
```

charettes replied:
Agreed and from a quick test it looks like multiple browsers sometimes dead lock the single threaded server. I have an other idea in mind which I'll share in few hours.

akki commented:
Further work by @charettes over here - https://github.com/django/django/compare/master...charettes:ticket-25364

akki commented:
buildbot, test on selenium.

charettes commented:
Looks like the test hanged :/

timgraham commented on tests/view_tests/tests/test_i18n.py:
```diff
@@ -3,19 +3,17 @@
-    override_settings, tag,
+    override_settings,
 )
+from django.test.selenium import SeleniumTestCaseMixin
```
`SeleniumTestCase`

charettes commented on tests/runtests.py:
```diff
@@ -233,6 +235,33 @@ def actual_test_processes(parallel):
+        except Exception:
+            raise RuntimeError(
+                'Selenium specifications not valid or corresponding WebDriver not installed: %s' % browser
+            )
```
This function is not used anymore.

akki commented:
Hi
This branch is not up to date, I was thinking maybe we should figure out how to make the tests pass on Jenkins and I could do the polishing after that (once we agree on a solution for the bigger problem).
Or do we plan to merge this PR anyways, as Jenkins is unstable with selenium tests ? In that case I will finish this off as soon as I can.

timgraham commented:
Simon continued this in https://github.com/django/django/pull/6272. Looks good to me pending a few cosmetic issues and additional comments. Thanks!
